### DISCUSSION TITLE
How to convert bool/str/float to float?

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3965] [Q&A] [answered]
Asked a few people, read many pages in The Book, but don’t get it: I need to be backwards compatible with a metadata variable that used to be boolean or boolean in a string, and is now a float value.

In Python, that was easy, but how to do in Liquidsoap? I didn’t even find something like `type` or `typeof` so I could ask LS for the variable type and then use an appropriate function.

Python code:
```
# Need to handle pre-version 3.0.0 `liq_blankskip`: was bool, is now float`
def float_blankskip(v):
    if isinstance(v, bool):
        return float(v) * args.blankskip  # True=1, False=0
    elif isinstance(v, str):
        try:
            return float(v)
        except ValueError:
            if v.lower() == 'true':
                return args.blankskip
            else:
                return 0.0
    else:
        return v
```

`args.blankskip` is a float # seconds here.

Any guru there who can help me with this?

**Comment by vitoyucepi** [comment, marked as answer]
If I were you, then I would convert everything to string, because every other type has a convenient to string conversion and then parse the string.
```ruby
def something_to_float(value)
  value_string = string.case(string(value))
  possible_float =
    try
      float_of_string(value_string)
    catch _ do
      null()
    end
  possible_bool =
    try
      bool_of_string(value_string) ? 1. : 0.
    catch _ do
      null()
    end
  possible_float ?? possible_bool ?? 0.
end
```

<details>
<summary>Tests</summary>

```ruby
a = true
b = false
c = "true"
d = "false"
e = "True"
f = "False"
g = "TRUE"
h = "FALSE"
i = 1.
j = -1.
k = "1."
l = "-1."
m = 1
n = -1
o = "1"
p = "-1"

def test(a)
  print(a)
  b = string.case(string(a))
  print(b)
  c =
    try
      float_of_string(b)
    catch _ do
      null()
    end
  d =
    try
      bool_of_string(b) ? 1. : 0.
    catch _ do
      null()
    end
  e = c ?? d
  print(e)
  print("")
end

test(a)
test(b)
test(c)
test(d)
test(e)
test(f)
test(g)
test(h)
test(i)
test(j)
test(k)
test(l)
test(m)
test(n)
test(o)
test(p)
```

</details>

**Reply by Moonbase59** [reply]
Fantastic. Thanks for all the help! There’s much to learn… ;-)

**Reply by toots** [reply]
This looks great @vitoyucepi. Python, ruby and JS typically have dynamic type conversion which makes these kind of operations really straight forward. However, they also come with some issues, typically having stuff converted to `NaN` without knowing because some of the input data is not what you expect it to be.

Static typing, which is OCaml, liquidsoap but also C and C++, typically require a little more care to handle errors and exceptions explicitly.

**Reply by Moonbase59** [reply]
@vitoyucepi Trying this out, I get:

```
At autocue.cue_file.liq, line 219, char 34-36:
  possible_float ?? possible_bool ?? 0.

Error 2: Parse error
```

Is it possible we can’t use `??` twice in a row?

**Reply by Moonbase59** [reply]
Put parentheses around, and it worked ;-)
```
  (possible_float ?? possible_bool) ?? 0.
```
